subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+		}
+
+		/*
+		 * Take over the next chunk if HCI packet is
+		 * larger than USB_MAX_FS_BULK_MPS.
+		 */
+		net_buf_add_mem(hci_data->acl_buf, buf->data, buf->len);
+		LOG_INF("len %u, chunk %u", hci_data->acl_buf->len, buf->len);
+	}
+
+	if (hci_data->acl_buf != NULL && hci_data->acl_len == hci_data->acl_buf->len) {
+		net_buf_put(&bt_hci_rx_queue, hci_data->acl_buf);
+		hci_data->acl_buf = NULL;
+		hci_data->acl_len = 0;
+	}
+
+restart_out_transfer:
+	/* TODO: add function to reset transfer buffer and allow to reuse it */
+	net_buf_unref(buf);
+	atomic_clear_bit(&hci_data->state, BT_HCI_ACL_RX_ENGAGED);
+	return bt_hci_acl_out_start(c_nd);
+}
+
+static int bt_hci_request(struct usbd_class_node *const c_nd,
+			  struct net_buf *buf, int err)
+{
+	struct usbd_contex *uds_ctx = c_nd->data->uds_ctx;
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == bt_hci_get_bulk_out(c_nd)) {
+		return bt_hci_acl_out_cb(c_nd, buf, err);
+	}
+
+	if (bi->ep == bt_hci_get_bulk_in(c_nd) || bi->ep == bt_hci_get_int_in(c_nd)) {
+		k_sem_give(&hci_data->sync_sem);
+		return 0;
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static void bt_hci_update(struct usbd_class_node *const c_nd,
+			  uint8_t iface, uint8_t alternate)
+{
+	LOG_DBG("New configuration, interface %u alternate %u",
+		iface, alternate);
+}
+
+static void bt_hci_enable(struct usbd_class_node *const c_nd)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+
+	atomic_set_bit(&hci_data->state, BT_HCI_CLASS_ENABLED);
+	LOG_INF("Configuration enabled");
+	if (bt_hci_acl_out_start(c_nd)) {
+		LOG_ERR("Failed to start ACL OUT transfer");
+	}
+}
+
+static void bt_hci_disable(struct usbd_class_node *const c_nd)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+
+	atomic_clear_bit(&hci_data->state, BT_HCI_CLASS_ENABLED);
+	LOG_INF("Configuration disabled");
+}
+
+static int bt_hci_ctd(struct usbd_class_node *const c_nd,
+		      const struct usb_setup_packet *const setup,
+		      const struct net_buf *const buf)
+{
+	struct net_buf *cmd_buf;
+
+	/* We expect host-to-device class request */
+	if (setup->RequestType.type != USB_REQTYPE_TYPE_CLASS) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("bmRequestType 0x%02x bRequest 0x%02x",
+		setup->bmRequestType, setup->bRequest);
+
+	cmd_buf = bt_buf_get_tx(BT_BUF_CMD, K_NO_WAIT, buf->data, buf->len);
+	if (!cmd_buf) {
+		LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	net_buf_put(&bt_hci_rx_queue, cmd_buf);
+
+	return 0;
+}
+
+static int bt_hci_init(struct usbd_class_node *const c_nd)
+{
+	return 0;
+}
+
+static struct usbd_class_api bt_hci_api = {
+	.request = bt_hci_request,
+	.update = bt_hci_update,
+	.enable = bt_hci_enable,
+	.disable = bt_hci_disable,
+	.control_to_dev = bt_hci_ctd,
+	.init = bt_hci_init,
+};
+
+#define BT_HCI_DESCRIPTOR_DEFINE(n)						\
+static struct usbd_bt_hci_desc bt_hci_desc_##n = {				\
+	.if0 = {								\
```

Comment:
Why doesn't this interface have a `bEndpointAddress`?

Reply:
Because it is interface descriptor and not an endpoint descriptor. All the endpoints belong to this interface. Notice that there is `bNumEndpoints` here with value 3.